exporter/signalfxexporter/translation/delta_translator.go

```diff
@@ -0,0 +1,103 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translation
+
+import (
+	"github.com/gogo/protobuf/proto"
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf/model"
+)
+
+type deltaTranslator struct {
+	prevPts *ttlMap
+}
+
+func newDeltaTranslator(ttl int64) *deltaTranslator {
+	sweepIntervalSeconds := ttl / 2
+	if sweepIntervalSeconds == 0 {
+		sweepIntervalSeconds = 1
+	}
+	m := newTTLMap(ttl, sweepIntervalSeconds)
+	m.start()
+	return &deltaTranslator{prevPts: m}
+}
+
+func (t *deltaTranslator) translate(pts []*sfxpb.DataPoint, tr Rule) []*sfxpb.DataPoint {
+	for _, currPt := range pts {
+		deltaMetricName, ok := tr.Mapping[currPt.Metric]
+		if !ok {
+			// only metrics defined in Rule.Mapping get translated
+			continue
+		}
+		deltaPt := t.deltaPt(deltaMetricName, currPt)
+		if deltaPt == nil {
+			continue
+		}
+		pts = append(pts, deltaPt)
+	}
+	return pts
+}
+
+func (t *deltaTranslator) deltaPt(deltaMetricName string, currPt *sfxpb.DataPoint) *sfxpb.DataPoint {
+	// check if we have a previous point for this metric + dimensions
+	dimKey := stringifyDimensions(currPt.Dimensions, nil)
+	fullKey := currPt.Metric + ":" + dimKey
+	v := t.prevPts.get(fullKey)
+	t.prevPts.put(fullKey, currPt)
+	if v == nil {
+		// no previous point, so we can't calculate a delta
+		return nil
+	}
+	prevPt := v.(*sfxpb.DataPoint)
+	var deltaPt *sfxpb.DataPoint
+	if currPt.Value.DoubleValue != nil && prevPt.Value.DoubleValue != nil {
+		deltaPt = doubleDeltaPt(currPt, prevPt, deltaMetricName)
+	} else if currPt.Value.IntValue != nil && prevPt.Value.IntValue != nil {
+		deltaPt = intDeltaPt(currPt, prevPt, deltaMetricName)
+	} else {
+		return nil
+	}
+	return deltaPt
+}
+
+func doubleDeltaPt(currPt *sfxpb.DataPoint, prevPt *sfxpb.DataPoint, deltaMetricName string) *sfxpb.DataPoint {
+	delta := *currPt.Value.DoubleValue - *prevPt.Value.DoubleValue
+	if delta < 0 {
+		// assume a reset, emit the current value
+		delta = *currPt.Value.DoubleValue
+	}
+	deltaPt := basePt(currPt, deltaMetricName)
+	*deltaPt.Value.DoubleValue = delta
+	return deltaPt
+}
+
+func intDeltaPt(currPt *sfxpb.DataPoint, prevPt *sfxpb.DataPoint, deltaMetricName string) *sfxpb.DataPoint {
+	delta := *currPt.Value.IntValue - *prevPt.Value.IntValue
+	if delta < 0 {
+		// assume a reset, emit the current value
+		delta = *currPt.Value.IntValue
+	}
+	deltaPt := basePt(currPt, deltaMetricName)
+	*deltaPt.Value.IntValue = delta
+	return deltaPt
+}
+
+var cumulativeCounterType = sfxpb.MetricType_CUMULATIVE_COUNTER
+
+func basePt(currPt *sfxpb.DataPoint, deltaMetricName string) *sfxpb.DataPoint {
+	deltaPt := proto.Clone(currPt).(*sfxpb.DataPoint)
+	deltaPt.Metric = deltaMetricName
+	deltaPt.MetricType = &cumulativeCounterType
```

Comment:
Why it's `sfxpb.MetricType_CUMULATIVE_COUNTER` ? It's supposed to be gauge instead, right?

Reply:
Hmm, the doc in model.pb.go for cumulative counter says "Tracks a value that increases over time, where only the difference is important." which seems like the right choice since we're emitting positive deltas?

Reply:
No deltas are not cumulative counters.  A cumulative counter is one that sends the latest full value of the total.  You probably want COUNTER.

Reply:
I don’t think counter makes sense for deltas. deltas are going up and down, that a definition of gauge. I see signalfx-agent uses gauge for cpu.utilization as well